src/cert-enroll/cert-enroll.in

```diff
@@ -0,0 +1,359 @@
+  then
+    echo "Error: generation of $size bit $KEYTYPE private key failed"
+    exit 1
+  fi
+  chmod 600 $1
+  echo "  generated $size bit $KEYTYPE private key '$1'"
+}
+
+function gen_cert_request()
+{
+  $PKI --req --in "$1/$HOSTKEY" --type $type \
+             --dn "$DN" $SAN1 $SAN2 $SAN3 $SAN4 $SAN5 \
+             --profile $PROFILE --outform pem > "$1/$CERTREQ"
+              
+  if [ $? -ne 0 -o ! -s $1 ]
+  then
+    echo "Error: generation of PKCS#10 certificate request failed"
+    exit 1
+  fi
+  chmod 600 $1
+  echo "  generated PKCS#10 certificate request"
+}
+
+function get_ca_certs()
+{
+  cd $1
+  if [ $CA_PROTOCOL == "EST" ]
+  then
+    $PKI --estca --url $EST_URL --cacert $TLSROOTCA --caout $CAOUT \
+                 --outform pem --force
+    if [ "$?" -ne 0 -o ! -s $ROOTCA -o ! -s $SUBCA ]
+    then
+      echo "Error: download of CA certificates via EST failed"
+      exit 1
+    fi
+    echo "  downloaded CA certificates via EST"
+  else
+    $PKI --scepca --url $SCEP_URL --caout $CAOUT --raout $RAOUT \
+                  --outform pem --force
+    if [ "$?" -ne 0 -o ! -s $ROOTCA -o ! -s $SUBCA -o ! -s $RACERT ]
+    then
+      echo "Error: download of CA or RA certificates via SCEP failed"
+      exit 1
+    fi
+    echo "  downloaded CA and RA certificates via SCEP"
+  fi
+  cd $CERTDIR
+}
+
+function check_ca_certs()
+{
+  get_ca_certs "$CERTDIR/new"
+
+  ROOTCA_CHANGED="0"
+  cmp -s $ROOTCA new/$ROOTCA
+  if [ "$?" -ne 0 ]
+  then
+    ROOTCA_CHANGED="1"
+    echo "Warning: '$ROOTCA' has changed"
+    mv $ROOTCA old
+    mv new/$ROOTCA .
+  fi
+
+  SUBCA_CHANGED="0"
+  cmp -s $SUBCA new/$SUBCA
+  if [ "$?" -ne 0 ]
+  then
+    SUBCA_CHANGED="1"
+    echo "Warning: '$SUBCA' has changed"
+    mv $SUBCA old
+    mv new/$SUBCA .
+  fi
+
+  if [ $CA_PROTOCOL == "SCEP" ]
+  then
+    mv new/$RACERT .
+  fi
+
+  if [ $ROOTCA_CHANGED -eq 0 -a $SUBCA_CHANGED -eq 0 ]
+  then
+    echo "Ok: '$ROOTCA' and '$SUBCA' are unchanged"
+    rm new/$ROOTCA new/$SUBCA
+    return 0
+  else
+    return 1
+  fi
+}
+
+function install_certs()
+{
+  for script in $INSTALL_SCRIPT_DIR/*
+  do
+    echo "  executing '$script'"
+    KEYTYPE="$KEYTYPE" CERTDIR="$CERTDIR" HOSTKEY="$HOSTKEY" \
+    HOSTCERT="$HOSTCERT" ROOTCA="$ROOTCA" SUBCA="$SUBCA" \
+    OLDROOTCA="$OLDROOTCA" OLDSUBCA="$OLDSUBCA" \
+    /bin/bash $script
+    if [ "$?" -ne 0 ]
+    then
+      echo "Error: executing '$script' failed"
+    fi
+  done
+}
+
+##############################################################################
+# SCEP certificate enrollment protocol requires RSA
+#
+if [ $PROTOCOL == "SCEP" -a $KEYTYPE == "ECDSA" ]
+then
+  echo "  the SCEP protocol does not support ECDSA keys, switched to RSA key"
+  KEYTYPE="RSA"
+fi
+
+##############################################################################
+# Select key size
+#
+if [ $KEYTYPE == "RSA" ]
+then
+  type="rsa"
+  size=$RSA_SIZE
+else
+  type="ecdsa"
+  size=$ECDSA_SIZE 
+fi 
```

Comment:
Changed to a `case` statement adding `ED25519`, `ED488`  cases and  as a default an `Error` message followed by `exit 1` for unknown key types. When we add new key types to `public_key.h` the `cert-enrol` script will be enhanced.

Reply:
OK. I guess since the script is not in its own repository and relies on support for new key types by `pki` anyway this is fine (one problem could be that we forget to update the script whenever new key types are added to `pki/public_key.h`, a generic fallback would at least offer an option to use such key types without having to patch the script).

Reply:
Support of other key types implemented. Unfortunately OpenXPKI doesn't support `ed25519` and `ed448` so that complete testing is not possible.